File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (!root) {
            return result;
        }
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            vector<int> nodes_at_level;
            const int q_size = q.size();
            for (int i = 0; i < q_size; i++) {
                TreeNode* front = q.front();
                q.pop();
                nodes_at_level.push_back(front -> val);
                if (front -> left) q.push(front -> left);
                if (front -> right) q.push(front -> right);
            }
            result.push_back(nodes_at_level);
        }

        return result;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def _dfs(self, node: Optional[TreeNode], k: int):
        if node:
            self._dfs(node.left, k)
            # process the node.
            self.node_number += 1
            if self.node_number == k:
                self.answer = node.val
            self._dfs(node.right, k)
        
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.answer = -1
        self.node_number = 0
        self._dfs(root, k)
        return self.answer

